crates/common_utils/src/pii.rs

```diff
@@ -88,25 +100,73 @@ where
 }
 
 /// Email address
-#[derive(Debug)]
-pub struct Email;
+#[derive(
+    serde::Serialize,
+    serde::Deserialize,
+    Debug,
+    Clone,
+    PartialEq,
+    Eq,
+    Default,
+    Queryable,
+    AsExpression,
+)]
+#[diesel(sql_type = diesel::sql_types::Text)]
+pub struct Email(Secret<String>);
+
+impl<DB> FromSql<sql_types::Text, DB> for Email
+where
+    DB: Backend,
+    String: FromSql<sql_types::Text, DB>,
+{
+    fn from_sql(bytes: backend::RawValue<'_, DB>) -> deserialize::Result<Self> {
+        let val = String::from_sql(bytes)?;
+        Ok(Self::from_str(val.as_str())?)
+    }
+}
+
+impl<DB> ToSql<sql_types::Text, DB> for Email
+where
+    DB: Backend,
+    String: ToSql<sql_types::Text, DB>,
+{
+    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, DB>) -> diesel::serialize::Result {
+        self.0.to_sql(out)
+    }
+}
+
+impl FromStr for Email {
+    type Err = ValidationError;
+
+    fn from_str(email: &str) -> Result<Self, Self::Err> {
+        if email.eq(REDACTED) {
+            return Ok(Self(Secret::new(email.to_string())));
+        }
+        match validate_email(email) {
+            Ok(_) => {
+                let secret: Secret<String> = Secret::new(email.to_string());
+                Ok(Self(secret))
+            }
+            Err(_) => Err(ValidationError::InvalidValue {
+                message: "Invalid email address format".into(),
+            }),
+        }
+    }
+}
 
 impl<T> Strategy<T> for Email
 where
-    T: AsRef<str>,
+    T: AsRef<str> + std::fmt::Debug,
 {
     fn fmt(val: &T, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         let val_str: &str = val.as_ref();
-        let is_valid = validate_email(val_str);
 
-        if is_valid.is_err() {
-            return WithType::fmt(val, f);
-        }
-
-        if let Some((a, b)) = val_str.split_once('@') {
-            write!(f, "{}@{}", "*".repeat(a.len()), b)
-        } else {
-            WithType::fmt(val, f)
+        match Self::from_str(val_str) {
+            Ok(_email) => match val_str.split_once('@') {
+                Some((a, b)) => write!(f, "{}@{}", "*".repeat(a.len()), b),
+                None => WithType::fmt(val, f),
+            },
+            Err(..) => WithType::fmt(val, f),
         }
     }
 }
```

Comment:
Hey, there seems to be a slight misunderstanding regarding the purpose of `Strategy<T>` and how it's supposed to work with a `Secret<D, S>`. Please refer to my comment [here](https://github.com/juspay/hyperswitch/pull/889#discussion_r1168201823).

To summarize, a `Secret<Data, Strategy>` by definition differentiates between data and a masking strategy, therefore a type (like `Email` here) that both, stores data and implements `Strategy<T>` is unintended.
Also even though `Email` implements `Strategy<T>` here, it's never used to mask the actual email in the definition (`struct Email(Secret<String>)`, no strategy provided to the secret). My suggestion would be to create a separate unit struct, implement `Strategy<T>` on said struct, and define this struct type as a strategy to the secret inside `Email`.

Reply:
Done